builtin/providers/aws/resource_aws_api_gateway_api_key.go

```diff
@@ -0,0 +1,225 @@
+				Required: true,
+			},
+
+			"enabled": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"stage_key": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"api_id": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"stage_name": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsApiGatewayStageKeys(d *schema.ResourceData) []*apigateway.StageKey {
+	var stageKeys []*apigateway.StageKey
+
+	if stageKeyData, ok := d.GetOk("stage_key"); ok {
+		params := stageKeyData.(*schema.Set).List()
+		for k := range params {
+			data := params[k].(map[string]interface{})
+			stageKeys = append(stageKeys, &apigateway.StageKey{
+				RestApiId: aws.String(data["api_id"].(string)),
+				StageName: aws.String(data["stage_name"].(string)),
+			})
+		}
+	}
+
+	return stageKeys
+}
+
+func resourceAwsApiGatewayApiKeyCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+	log.Printf("[DEBUG] Creating API Gateway API Key")
+
+	apiKey, err := conn.CreateApiKey(&apigateway.CreateApiKeyInput{
+		Name:        aws.String(d.Get("name").(string)),
+		Description: aws.String(d.Get("description").(string)),
+		Enabled:     aws.Bool(d.Get("enabled").(bool)),
+		StageKeys:   resourceAwsApiGatewayStageKeys(d),
+	})
+	if err != nil {
+		return fmt.Errorf("Error creating API Gateway: %s", err)
+	}
+
+	d.SetId(*(*apiKey).Id)
+
+	return resourceAwsApiGatewayApiKeyRead(d, meta)
+}
+
+func resourceAwsApiGatewayApiKeyRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+	log.Printf("[DEBUG] Reading API Gateway API Key: %s", d.Id())
+
+	apiKey, err := conn.GetApiKey(&apigateway.GetApiKeyInput{
+		ApiKey: aws.String(d.Id()),
+	})
+	if err != nil {
+		return err
+	}
+
+	d.Set("name", *apiKey.Name)
+	d.Set("description", *apiKey.Description)
+	d.Set("enabled", *apiKey.Enabled)
+
+	return nil
+}
+
+func resourceAwsApiGatewayApiKeyUpdateOperations(d *schema.ResourceData) []*apigateway.PatchOperation {
+	operations := make([]*apigateway.PatchOperation, 0)
+	if d.HasChange("enabled") {
+		isEnabled := "false"
+		if d.Get("enabled").(bool) {
+			isEnabled = "true"
+		}
+		operations = append(operations, &apigateway.PatchOperation{
+			Op:    aws.String("replace"),
+			Path:  aws.String("/enabled"),
+			Value: aws.String(isEnabled),
+		})
+	}
+
+	if d.HasChange("description") {
+		operations = append(operations, &apigateway.PatchOperation{
+			Op:    aws.String("replace"),
+			Path:  aws.String("/description"),
+			Value: aws.String(d.Get("description").(string)),
+		})
+	}
+
+	if d.HasChange("stage_key") {
+		prev, curr := d.GetChange("stage_key")
+		prevList := prev.(*schema.Set).List()
```

Comment:
Would you mind decoupling this logic for detection add/removal into a separate function and moving it into `builtin/providers/aws/structure.go`?